lib/ReplaceOpenCLBuiltinPass.cpp

```diff
@@ -660,6 +675,191 @@ Type *ReplaceOpenCLBuiltinPass::GetPairStruct(Type *type) {
   return new_struct;
 }
 
+bool ReplaceOpenCLBuiltinPass::replaceWaitGroupEvents(Function &F) {
+  /* Simple implementation for wait_group_events to avoid dealing with the event
+   * list:
+   *
+   * void wait_group_events(int num_events, event_t *event_list) {
+   *   barrier(CLK_LOCAL_MEM_FENCE);
+   * }
+   *
+   */
+
+  enum {
+    CLK_LOCAL_MEM_FENCE = 0x01,
+    CLK_GLOBAL_MEM_FENCE = 0x02,
+    CLK_IMAGE_MEM_FENCE = 0x04
+  };
+
+  return replaceCallsWithValue(F, [](CallInst *CI) {
+    IRBuilder<> Builder(CI);
+
+    const auto ConstantScopeWorkgroup = Builder.getInt32(spv::ScopeWorkgroup);
+    const auto MemorySemanticsWorkgroup = BinaryOperator::Create(
+        Instruction::Shl, Builder.getInt32(CLK_LOCAL_MEM_FENCE),
+        Builder.getInt32(clz(spv::MemorySemanticsWorkgroupMemoryMask) -
+                         clz(CLK_LOCAL_MEM_FENCE)),
+        "", CI);
+
+    return clspv::InsertSPIRVOp(CI, spv::OpControlBarrier,
+                                {Attribute::NoDuplicate, Attribute::Convergent},
+                                Builder.getVoidTy(),
+                                {ConstantScopeWorkgroup, ConstantScopeWorkgroup,
+                                 MemorySemanticsWorkgroup});
+  });
+}
+
+Value *ReplaceOpenCLBuiltinPass::replaceAsyncWorkGroupCopies(
+    Module &M, CallInst *CI, Value *Dst, Value *Src, Value *NumGentypes,
+    Value *Stride, Value *Event) {
+  /*
+   * event_t *async_work_group_strided_copy(T *dst, T *src, size_t num_gentypes,
+   *                                        size_t stride, event_t event) {
+   *   size_t start_id = ((get_local_id(2) * get_local_size(1))
+   *                     + get_local_id(1)) * get_local_size(0)
+   *                     + get_local_id(0);
+   *   size_t incr = get_local_size(0) * get_local_size(1) * get_local_size(2);
+   *   for (size_t it = start_id; it < num_gentypes; it += incr) {
+   *     dst[it] = src[it * stride];
+   *   }
```

Comment:
You are talking of this kind of implementation I suppose
```
    size_t nb_element_per_wi = num_gentypes / group_size;
    size_t remainder = num_gentypes % group_size; // OR num_gentypes - nb_element_per_wi * group_size
    size_t first_element = id * nb_element_per_wi + ((id < remainder) ? id : remainder);
    size_t last_element = first_element + nb_element_per_wi + ((id < remainder) ? 1 : 0);
    for (size_t i = first_element; i < last_element; i++) {
        dst[i] = src[i];
    }
```
I agree it could be more efficient. I will try it to have a look, but my first idea was to go for something simpler as this feature is mostly not used. Just enough to have the right behavior felt enough to me. I will try to implement this as well